# Project 2 - Ames Housing Data and Kaggle Challenge (01_Data_Cleaning_Train_Set)

## Background Summary

Provided with Ames housing data, there many contributing features that could possibly be of great impact or less in influencing housing prices. With this insight, house buyers could look into the more prominent and proper type of feature to make a judgement to assist propery companies to make a better rated pricing for houses. With this awareness of selectivity in the contributing predictive features, a proposed model will be constructed to allow an economical or a competitive pricing to be improved for potential house buyers and sellers.  

## Problem Statement

There are many feature predictors based on which may potentially greatly affect the sales price of houses in Ames. The focus for this project is to streamline it down from all the feature predictors to the top 5 features that may greatly be of a potential impact to the sales price of housing in Ames.

## Data Dictionary

| No | Features/ ID        | Type        | Description                                                                                    |
|----|---------------------|-------------|-----------------------------------------------------------------------------------------------|
| 0  |   id                | int         | ID number                                                                                      |
| 1  |   pid               | int         |  Parcel identification number  - can be used with city web site for parcel review.             |
| 2  |   ms_subclass       | int         | Identifies the type of dwelling involved in the sale.                                        |
| 3  |   ms_zoning         | categorical | Identifies the general zoning classification of the sale.                                    |
| 4  |   lot_frontage      | float       | Linear feet of street connected to property                                                    |
| 5  |   lot_area          | int         | Lot size in square feet                                                                        |
| 6  |   street            | categorical | Type of road access to property                                                                |
| 7  |   lot_shape         | categorical | General shape of property                                                                      |
| 8  |   land_contour      | categorical | Flatness of the property                                                                      |
| 9  |   utilities         | categorical | Type of utilities available                                                                    |
| 10 |   lot_config        | categorical | Lot configuration                                                                              |
| 11 |   land_slope        | categorical | Slope of property                                                                              |
| 12 |   neighborhood      | categorical | Physical locations within Ames city limits (map available)                                   |
| 13 |   condition_1       | categorical | Proximity to various conditions                                                                |
| 14 |   condition_2       | categorical | Proximity to various conditions (if more than one is present)                                  |         
| 15 |   bldg_type         | categorical | Type of dwelling                                                                              |
| 16 |   house_style       | categorical | Style of dwelling                                                                              |
| 17 |   overall_qual      | int         | Rates the overall material and finish of the house                                        |
| 18 |   overall_cond      | int         | Rates the overall condition of the house                                                      |
| 19 |   year_built        | int         | Original construction date                                                                    |
| 20 |   year_remod/add    | int         | Remodel date (same as construction date if no remodeling or additions)                     |
| 21 |   roof_style        | categorical | Type of roof                                                                                  |
| 23 |   exterior_1st      | categorical | Exterior covering on house                                                                    |
| 24 |   exterior_2nd      | categorical | Exterior covering on house (if more than one material)                                    |
| 25 |   mas_vnr_type      | categorical | Masonry veneer type                                                                            |
| 26 |   mas_vnr_area      | float       | Masonry veneer area in square feet                                                            |
| 27 |   exter_qual        | categorical | Evaluates the quality of the material on the exterior                                     |
| 28 |   exter_cond        | categorical | Evaluates the present condition of the material on the exterior                              |
| 29 |   foundation        | categorical | Type of foundation                                                                            |
| 30 |   bsmt_qual         | categorical | Evaluates the height of the basement                                                          |
| 31 |   bsmt_cond         | categorical | Evaluates the general condition of the basement                                                |
| 32 |   bsmt_exposure     | categorical | Refers to walkout or garden level walls                                                        |
| 33 |   bsmtfin_type_1    | categorical | Rating of basement finished area                                                              |
| 34 |   bsmtfin_sf_1      | int         | Type 1 finished square feet                                                                    |
| 35 |   bsmtfin_type_2    | categorical | Rating of basement finished area (if multiple types)                                       |
| 36 |   bsmtfin_sf_2      | int         | Type 2 finished square feet                                                                    |
| 37 |   bsmt_unf_sf       | int         | Unfinished square feet of basement area                                                        |
| 38 |   total_bsmt_sf     | int         | Total square feet of basement area                                                            |
| 39 |   heating           | categorical | Type of heating                                                                                |
| 40 |   heating_qc        | categorical | Heating quality and condition                                                                  |
| 41 |   central_air       | categorical | Central air conditioning                                                                      |
| 42 |   electrical        | categorical | Electrical system                                                                              |
| 43 |   1st_flr_sf        | int         | First Floor square feet                                                                        |
| 44 |   2nd_flr_sf        | int         | Second floor square feet                                                                      |
| 45 |   low_qual_fin_sf   | int         | Low quality finished square feet (all floors)                                                  |
| 46 |   gr_liv_area       | int         | Above grade (ground) living area square feet                                                  |
| 47 |   bsmt_full_bath    | int         | Basement full bathrooms                                                                        |
| 48 |   bsmt_half_bath    | int         | Basement half bathrooms                                                                        |
| 49 |   full_bath         | int         | Full bathrooms above grade                                                                    |
| 50 |   half_bath         | int         | Half baths above grade                                                                        |
| 51 |   bedroom_abvgr     | int         | Bedrooms above grade (does NOT include basement bedrooms)                                    |
| 52 |   kitchen_abvgr     | int         | Kitchens above grade                                                                          |
| 53 |   kitchen_qual      | categorical | Kitchen quality                                                                                |
| 54 |   totrms_abvgrd     | int         | Total rooms above grade (does not include bathrooms)                                   |
| 55 |   functional        | categorical | Home functionality (Assume typical unless deductions are warranted)                    |
| 56 |   fireplaces        | int         | Number of fireplaces                                                                          |
| 57 |   fireplace_qu      | categorical | Fireplace quality                                                                              |
| 58 |   garage_type       | categorical | Garage location                                                                                |
| 59 |   garage_yr_blt     | float       | Year garage was built                                                                          |
| 60 |   garage_finish     | categorical | Interior finish of the garage                                                                  |
| 61 |   garage_cars       | int         | Size of garage in car capacity                                                                |
| 62 |   garage_area       | int         | Size of garage in square feet                                                                  |
| 63 |   garage_qual       | categorical | Garage quality                                                                                |
| 64 |   garage_cond       | categorical | Garage condition                                                                              |
| 65 |   paved_drive       | categorical | Paved driveway                                                                                |
| 66 |   wood_deck_sf      | int         | Wood deck area in square feet                                                                  |
| 67 |   open_porch_sf     | int         | Open porch area in square feet                                                                |
| 68 |   enclosed_porch    | int         | Enclosed porch area in square feet                                                            |
| 69 |   3ssn_porch        | int         | Three season porch area in square feet                                                        |
| 70 |   screen_porch      | int         | Screen porch area in square feet                                                              |
| 71 |   pool_area         | int         | Pool area in square feet                                                                      |
| 72 |   pool_qc           | categorical | Pool quality                                                                                  |
| 73 |   fence             | categorical | Fence quality                                                                                  |
| 74 |   misc_feature      | categorical | Miscellaneous feature not covered in other categories                                   |
| 75 |   misc_val          | int         | Value of miscellaneous feature                                                                |
| 76 |   mo_sold           | int         | Month Sold (MM)                                                                                |
| 77 |   yr_sold           | int         | Year Sold (YYYY)                                                                              |
| 78 |   sale_type         | categorical | Type of sale                                                                                  |
| 79 |   saleprice         | int         | Condition of sale                                                                              |
| 80 |   alley             | categorical | Type of alley access to property                                                              |
| 81 |   prop_age          | int         | Age of Property in years                                                                      |
| 81 |   remod_age         | int         | Age of remodel/add in years                                                                    |



### Data Description

train.csv -- this data contains all of the training data for your model.
The target variable (SalePrice) is removed from the test set!

test.csv -- this data contains the test data for your model. You will feed this data into your regression model to make predictions.

sample_sub_reg.csv -- An example of a correctly formatted submission for this challenge (with a random number provided as predictions for SalePrice. Please ensure that your submission to Kaggle matches this format.

## Libaries Imports

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression, LassoCV, RidgeCV
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
from sklearn.model_selection import train_test_split, cross_val_score, cross_val_predict
from sklearn.metrics import r2_score

%matplotlib inline 

## Data Imports

In [2]:
pd.options.display.max_columns = None
train = pd.read_csv('../datasets/train.csv')

In [3]:
sample_sub_reg = pd.read_csv('../datasets/sample_sub_reg.csv')

In [4]:
sample_sub_reg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 878 entries, 0 to 877
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Id         878 non-null    int64  
 1   SalePrice  878 non-null    float64
dtypes: float64(1), int64(1)
memory usage: 13.8 KB


## Data Cleaning - training set

In [5]:
train.shape

(2051, 81)

In [6]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2051 entries, 0 to 2050
Data columns (total 81 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               2051 non-null   int64  
 1   PID              2051 non-null   int64  
 2   MS SubClass      2051 non-null   int64  
 3   MS Zoning        2051 non-null   object 
 4   Lot Frontage     1721 non-null   float64
 5   Lot Area         2051 non-null   int64  
 6   Street           2051 non-null   object 
 7   Alley            140 non-null    object 
 8   Lot Shape        2051 non-null   object 
 9   Land Contour     2051 non-null   object 
 10  Utilities        2051 non-null   object 
 11  Lot Config       2051 non-null   object 
 12  Land Slope       2051 non-null   object 
 13  Neighborhood     2051 non-null   object 
 14  Condition 1      2051 non-null   object 
 15  Condition 2      2051 non-null   object 
 16  Bldg Type        2051 non-null   object 
 17  House Style   

##### Checking if Id and PID are unique

In [7]:
print('Total unique rows for Id:', len(train['Id'].unique()))
print('Total unique rows for PID:', len(train['PID'].unique()))

Total unique rows for Id: 2051
Total unique rows for PID: 2051


##### Setting column names to lower case and replacing spacing with a '_'

In [8]:
train.columns = train.columns.str.lower()
train.columns = train.columns.str.replace(' ', '_')
train.head()

id        pid  ms_subclass ms_zoning  lot_frontage  lot_area street alley  \
0  109  533352170           60        RL           NaN     13517   Pave   NaN   
1  544  531379050           60        RL          43.0     11492   Pave   NaN   
2  153  535304180           20        RL          68.0      7922   Pave   NaN   
3  318  916386060           60        RL          73.0      9802   Pave   NaN   
4  255  906425045           50        RL          82.0     14235   Pave   NaN   

  lot_shape land_contour utilities lot_config land_slope neighborhood  \
0       IR1          Lvl    AllPub    CulDSac        Gtl       Sawyer   
1       IR1          Lvl    AllPub    CulDSac        Gtl      SawyerW   
2       Reg          Lvl    AllPub     Inside        Gtl        NAmes   
3       Reg          Lvl    AllPub     Inside        Gtl       Timber   
4       IR1          Lvl    AllPub     Inside        Gtl      SawyerW   

  condition_1 condition_2 bldg_type house_style  overall_qual  overall_cond  \
0        RRAe        Norm      1Fam      2Story             6             8   
1        Norm        Norm      1Fam      2Story             7             5   
2        Norm        Norm      1Fam      1Story             5             7   
3        Norm        Norm      1Fam      2Story             5             5   
4        Norm        Norm      1Fam      1.5Fin             6             8   

   year_built  year_remod/add roof_style roof_matl exterior_1st exterior_2nd  \
0        1976            2005      Gable   CompShg      HdBoard      Plywood   
1        1996            1997      Gable   CompShg      VinylSd      VinylSd   
2        1953            2007      Gable   CompShg      VinylSd      VinylSd   
3        2006            2007      Gable   CompShg      VinylSd      VinylSd   
4        1900            1993      Gable   CompShg      Wd Sdng      Plywood   

  mas_vnr_type  mas_vnr_area exter_qual exter_cond foundation bsmt_qual  \
0      BrkFace         289.0         Gd         TA     CBlock        TA   
1      BrkFace         132.0         Gd         TA      PConc        Gd   
2         None           0.0         TA         Gd     CBlock        TA   
3         None           0.0         TA         TA      PConc        Gd   
4         None           0.0         TA         TA      PConc        Fa   

  bsmt_cond bsmt_exposure bsmtfin_type_1  bsmtfin_sf_1 bsmtfin_type_2  \
0        TA            No            GLQ         533.0            Unf   
1        TA            No            GLQ         637.0            Unf   
2        TA            No            GLQ         731.0            Unf   
3        TA            No            Unf           0.0            Unf   
4        Gd            No            Unf           0.0            Unf   

   bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf heating heating_qc central_air  \
0           0.0        192.0          725.0    GasA         Ex           Y   
1           0.0        276.0          913.0    GasA         Ex           Y   
2           0.0        326.0         1057.0    GasA         TA           Y   
3           0.0        384.0          384.0    GasA         Gd           Y   
4           0.0        676.0          676.0    GasA         TA           Y   

  electrical  1st_flr_sf  2nd_flr_sf  low_qual_fin_sf  gr_liv_area  \
0      SBrkr         725         754                0         1479   
1      SBrkr         913        1209                0         2122   
2      SBrkr        1057           0                0         1057   
3      SBrkr         744         700                0         1444   
4      SBrkr         831         614                0         1445   

   bsmt_full_bath  bsmt_half_bath  full_bath  half_bath  bedroom_abvgr  \
0             0.0             0.0          2          1              3   
1             1.0             0.0          2          1              4   
2             1.0             0.0          1          0              3   
3             0.0             0.0          2  

##### Checking for numerical data type values either float or int

In [9]:
train.select_dtypes(np.number).head()

id        pid  ms_subclass  lot_frontage  lot_area  overall_qual  \
0  109  533352170           60           NaN     13517             6   
1  544  531379050           60          43.0     11492             7   
2  153  535304180           20          68.0      7922             5   
3  318  916386060           60          73.0      9802             5   
4  255  906425045           50          82.0     14235             6   

   overall_cond  year_built  year_remod/add  mas_vnr_area  bsmtfin_sf_1  \
0             8        1976            2005         289.0         533.0   
1             5        1996            1997         132.0         637.0   
2             7        1953            2007           0.0         731.0   
3             5        2006            2007           0.0           0.0   
4             8        1900            1993           0.0           0.0   

   bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf  1st_flr_sf  2nd_flr_sf  \
0           0.0        192.0          725.0         725         754   
1           0.0        276.0          913.0         913        1209   
2           0.0        326.0         1057.0        1057           0   
3           0.0        384.0          384.0         744         700   
4           0.0        676.0          676.0         831         614   

   low_qual_fin_sf  gr_liv_area  bsmt_full_bath  bsmt_half_bath  full_bath  \
0                0         1479             0.0             0.0          2   
1                0         2122             1.0             0.0          2   
2                0         1057             1.0             0.0          1   
3                0         1444             0.0             0.0          2   
4                0         1445             0.0             0.0          2   

   half_bath  bedroom_abvgr  kitchen_abvgr  totrms_abvgrd  fireplaces  \
0          1              3              1              6           0   
1          1              4              1              8           1   
2          0              3              1              5           0   
3          1              3              1              7           0   
4          0              3              1              6           0   

   garage_yr_blt  garage_cars  garage_area  wood_deck_sf  open_porch_sf  \
0         1976.0          2.0        475.0             0             44   
1         1997.0          2.0        559.0             0             74   
2         1953.0          1.0        246.0             0             52   
3         2007.0          2.0        400.0           100              0   
4         1957.0          2.0        484.0             0             59   

   enclosed_porch  3ssn_porch  screen_porch  pool_area  misc_val  mo_sold  \
0               0           0             0          0         0        3   
1               0           0             0          0         0        4   
2               0           0             0          0         0        1   
3               0           0             0          0         0        4   
4               0           0             0          0         0        3   

   yr_sold  saleprice  
0     2010     130500  
1     2009     220000  
2     2010     109000  
3     2010     174000  
4     2010     138500

In [145]:
# Assigning all variables that are numerical to num_column
num_column = [col for col in train.columns 
              if (train[col].dtype == 'int64' or train[col].dtype == float)]
num_column

['id',
 'pid',
 'ms_subclass',
 'lot_frontage',
 'lot_area',
 'overall_qual',
 'overall_cond',
 'year_built',
 'year_remod/add',
 'mas_vnr_area',
 'bsmtfin_sf_1',
 'bsmtfin_sf_2',
 'bsmt_unf_sf',
 'total_bsmt_sf',
 '1st_flr_sf',
 '2nd_flr_sf',
 'low_qual_fin_sf',
 'gr_liv_area',
 'bsmt_full_bath',
 'bsmt_half_bath',
 'full_bath',
 'half_bath',
 'bedroom_abvgr',
 'kitchen_abvgr',
 'totrms_abvgrd',
 'fireplaces',
 'garage_yr_blt',
 'garage_cars',
 'garage_area',
 'wood_deck_sf',
 'open_porch_sf',
 'enclosed_porch',
 '3ssn_porch',
 'screen_porch',
 'pool_area',
 'misc_val',
 'mo_sold',
 'yr_sold',
 'saleprice']

##### Listing out columns for sub categorising

In [11]:
train.columns

Index(['id', 'pid', 'ms_subclass', 'ms_zoning', 'lot_frontage', 'lot_area',
       'street', 'alley', 'lot_shape', 'land_contour', 'utilities',
       'lot_config', 'land_slope', 'neighborhood', 'condition_1',
       'condition_2', 'bldg_type', 'house_style', 'overall_qual',
       'overall_cond', 'year_built', 'year_remod/add', 'roof_style',
       'roof_matl', 'exterior_1st', 'exterior_2nd', 'mas_vnr_type',
       'mas_vnr_area', 'exter_qual', 'exter_cond', 'foundation', 'bsmt_qual',
       'bsmt_cond', 'bsmt_exposure', 'bsmtfin_type_1', 'bsmtfin_sf_1',
       'bsmtfin_type_2', 'bsmtfin_sf_2', 'bsmt_unf_sf', 'total_bsmt_sf',
       'heating', 'heating_qc', 'central_air', 'electrical', '1st_flr_sf',
       '2nd_flr_sf', 'low_qual_fin_sf', 'gr_liv_area', 'bsmt_full_bath',
       'bsmt_half_bath', 'full_bath', 'half_bath', 'bedroom_abvgr',
       'kitchen_abvgr', 'kitchen_qual', 'totrms_abvgrd', 'functional',
       'fireplaces', 'fireplace_qu', 'garage_type', 'garage_yr_blt',
       'g

##### Categorising columns into a sub categorised column for referencing
- __class_column:__  Class related field
- __date_column:__  Time line related field
- __ext_column:__  Exterior related field
- __int_column:__  Interior related field
- __bsmt_column:__  Basement related field
- __util_column:__  Utilities related field
- __sqft_column:__  Area related field
- __bath_column:__  Bathroom related field
- __sale_column:__  Sales related field
- __qc_column:__  Quailities or state of conditions related field
- __gar_column:__  Garage related field 

In [12]:
class_column = ['id', 'pid', 'ms_subclass', 'ms_zoning', 'bldg_type', 'house_style', 
             'overall_qual', 'overall_cond', 'lot_shape', 'land_contour', 
             'lot_config', 'land_slope', 'functional']

date_column = ['id', 'pid', 'year_built', 'year_remod/add', 'mo_sold', 'yr_sold']

ext_column = ['id', 'pid', 'street', 'alley', 'exterior_1st', 'exterior_2nd', 'mas_vnr_type','mas_vnr_area', 'exter_qual',
           'exter_cond', 'foundation', 'neighborhood','condition_1', 'condition_2', 'paved_drive', 'pool_qc','fence']

int_column = ['id', 'pid', 'bsmt_full_bath','bsmt_half_bath', 'full_bath', 'half_bath', 'bedroom_abvgr','kitchen_abvgr', 
           'kitchen_qual', 'totrms_abvgrd','bsmt_qual','bsmt_cond', 'bsmt_exposure', 'bsmtfin_type_1', 
           'bsmtfin_sf_1','bsmtfin_type_2', 'bsmtfin_sf_2', 'bsmt_unf_sf', 'total_bsmt_sf', 'fireplaces', 
           'fireplace_qu', 'garage_type', 'garage_yr_blt', 'garage_finish', 'garage_cars', 'garage_area', 
           'garage_qual','garage_cond']
bsmt_column = ['bsmt_qual', 'bsmt_cond', 'bsmt_exposure', 'bsmtfin_type_1', 'bsmtfin_sf_1','bsmtfin_type_2', 'bsmtfin_sf_2', 
            'bsmt_unf_sf','total_bsmt_sf', 'bsmt_full_bath','bsmt_half_bath']

util_column = ['id', 'pid', 'utilities', 'heating', 'heating_qc', 'central_air', 'electrical', 'misc_feature', 'misc_val']

sqft_column = ['id', 'pid', '2nd_flr_sf', 'low_qual_fin_sf', 'gr_liv_area', 'lot_area', 'wood_deck_sf', 'open_porch_sf', 
            'enclosed_porch', '3ssn_porch', 'screen_porch', 'pool_area']

bath_column = ['id', 'pid', 'bsmt_full_bath', 'bsmt_half_bath', 'full_bath', 'half_bath', 'bedroom_abvgr', 'kitchen_abvgr',
                 'kitchen_qual', 'totrms_abvgrd']

sale_column = ['id', 'pid', 'mo_sold', 'yr_sold', 'sale_type', 'saleprice']

qc_column = ['id', 'pid', 'overall_qual', 'overall_cond', 'exter_qual', 'exter_cond', 'pool_qc', 'kitchen_qual', 'bsmt_qual',
          'bsmt_cond', 'fireplace_qu', 'garage_qual','garage_cond', 'heating_qc']

gar_column = ['garage_type', 'garage_yr_blt', 'garage_finish', 'garage_cars', 'garage_area', 
           'garage_qual','garage_cond']

##### Checking for Duplicated values

In [13]:
print('Duplicated train values:', train.duplicated().sum())

Duplicated train values: 0


## Null Values

### Checking for Null values

In [14]:
train.isnull().sum().sort_values(ascending=False)[:29]

pool_qc           2042
misc_feature      1986
alley             1911
fence             1651
fireplace_qu      1000
lot_frontage       330
garage_finish      114
garage_cond        114
garage_qual        114
garage_yr_blt      114
garage_type        113
bsmt_exposure       58
bsmtfin_type_2      56
bsmtfin_type_1      55
bsmt_cond           55
bsmt_qual           55
mas_vnr_type        22
mas_vnr_area        22
bsmt_half_bath       2
bsmt_full_bath       2
garage_cars          1
garage_area          1
bsmt_unf_sf          1
bsmtfin_sf_2         1
total_bsmt_sf        1
bsmtfin_sf_1         1
overall_cond         0
exterior_2nd         0
exterior_1st         0
dtype: int64

### Handling Numerical Null Values

In [15]:
#Getting the total null values for the numerical columns in train dataset
train_null_num = pd.DataFrame(train[num_column].isnull().sum(), columns=['null_count'])
train_null_num = train_null_num[train_null_num['null_count'] > 0].sort_values('null_count', ascending=False).sort_index()
train_null_num

null_count
bsmt_full_bath           2
bsmt_half_bath           2
bsmt_unf_sf              1
bsmtfin_sf_1             1
bsmtfin_sf_2             1
garage_area              1
garage_cars              1
garage_yr_blt          114
lot_frontage           330
mas_vnr_area            22
total_bsmt_sf            1

##### Cleaning column bsmt_full_bath null values

In [16]:
# Identifying row of NaN values
train[train['bsmt_full_bath'].isna()]['bsmt_full_bath']

616    NaN
1327   NaN
Name: bsmt_full_bath, dtype: float64

In [17]:
# Setting NaN values to 0 
train.loc[train[train['bsmt_full_bath'].isna()].index, 'bsmt_full_bath'] = 0.0

In [18]:
# Verifying that the NaN values has been changed
train.groupby('bsmt_full_bath').count()['id'].sum()

2051

In [19]:
train.groupby('bsmt_full_bath').count()['id']

bsmt_full_bath
0.0    1202
1.0     824
2.0      23
3.0       2
Name: id, dtype: int64

##### Cleaning column bsmt_half_bath null values

In [20]:
# Identifying row of NaN values
train[train['bsmt_half_bath'].isna()]['bsmt_half_bath']

616    NaN
1327   NaN
Name: bsmt_half_bath, dtype: float64

In [21]:
# Setting NaN values to 0 
train.loc[train[train['bsmt_half_bath'].isna()].index, 'bsmt_half_bath'] = 0.0

In [22]:
# Verifying that the NaN values has been changed
train.groupby('bsmt_half_bath').count()['id'].sum()

2051

In [23]:
train.groupby('bsmt_half_bath').count()['id']

bsmt_half_bath
0.0    1925
1.0     122
2.0       4
Name: id, dtype: int64

##### Cleaning column bsmt_unf_sf, bsmtfin_sf_1, bsmtfin_sf_2  null values

In [24]:
# Identifying row of NaN values
train[train['bsmt_unf_sf'].isna()]['bsmt_unf_sf']

1327   NaN
Name: bsmt_unf_sf, dtype: float64

In [25]:
# Setting NaN values to 0 
train.loc[train[train['bsmt_unf_sf'].isna()].index, 'bsmt_unf_sf'] = 0.0

In [26]:
# Verifying that the NaN values have been changed
train.groupby('bsmt_unf_sf').count()['id'].sum()

2051

In [27]:
# Identifying row of NaN values
train[train['bsmtfin_sf_1'].isna()]['bsmtfin_sf_1']

1327   NaN
Name: bsmtfin_sf_1, dtype: float64

In [28]:
# Setting NaN values to 0 
train.loc[train[train['bsmtfin_sf_1'].isna()].index, 'bsmtfin_sf_1'] = 0.0

In [29]:
# Verifying that the NaN values have been changed
train.groupby('bsmtfin_sf_1').count()['id'].sum()

2051

In [30]:
# Identifying row of NaN values
train[train['bsmtfin_sf_2'].isna()]['bsmtfin_sf_2']

1327   NaN
Name: bsmtfin_sf_2, dtype: float64

In [31]:
# Setting NaN values to 0 
train.loc[train[train['bsmtfin_sf_2'].isna()].index, 'bsmtfin_sf_2'] = 0.0

In [32]:
# Verifying that the NaN values have been changed
train.groupby('bsmtfin_sf_2').count()['id'].sum()

2051

##### Cleaning column garage_area, garage_cars null values

In [33]:
train[train['garage_area'] == 0].head()

id        pid  ms_subclass ms_zoning  lot_frontage  lot_area street  \
28   2243  911103060          190   C (all)          75.0      8250   Pave   
53    330  923226250          160        RM          21.0      1476   Pave   
65   2278  923202134           20        RL          70.0      8402   Pave   
79   2235  910201050           50        RM          50.0      7288   Pave   
101  2084  905476170           30        RL          58.0      9098   Pave   

    alley lot_shape land_contour utilities lot_config land_slope neighborhood  \
28    NaN       Reg          Lvl    AllPub     Inside        Gtl       IDOTRR   
53    NaN       Reg          Lvl    AllPub     Inside        Gtl      MeadowV   
65    NaN       Reg          Lvl    AllPub     Inside        Gtl      Mitchel   
79    NaN       Reg          Lvl    AllPub     Inside        Gtl       IDOTRR   
101   NaN       IR1          Lvl    AllPub     Inside        Gtl      Edwards   

    condition_1 condition_2 bldg_type house_style  overall_qual  overall_cond  \
28        Feedr        Norm    2fmCon      2Story             5             6   
53         Norm        Norm     Twnhs      2Story             4             7   
65        Feedr        Norm      1Fam      1Story             5             5   
79       Artery        Norm      1Fam      1.5Fin             5             7   
101        Norm        Norm      1Fam      1Story             4             7   

     year_built  year_remod/add roof_style roof_matl exterior_1st  \
28         1895            2006      Gable   CompShg      VinylSd   
53         1970            1970      Gable   CompShg      CemntBd   
65         2007            2007      Gable   CompShg      VinylSd   
79         1925            2003      Gable   CompShg      VinylSd   
101        1920            2002      Gable   CompShg      Wd Sdng   

    exterior_2nd mas_vnr_type  mas_vnr_area exter_qual exter_cond foundation  \
28       VinylSd         None           0.0         TA         Gd     CBlock   
53       CmentBd         None           0.0         TA         TA     CBlock   
65       VinylSd         None           0.0         TA         TA      PConc   
79       VinylSd         None           0.0         Gd         Gd     BrkTil   
101      Wd Sdng         None           0.0         TA         TA     BrkTil   

    bsmt_qual bsmt_cond bsmt_exposure bsmtfin_type_1  bsmtfin_sf_1  \
28         TA        TA            No            Unf           0.0   
53         TA        TA            No            GLQ         176.0   
65         Gd        TA            No            ALQ         206.0   
79         TA        Po            No            Unf           0.0   
101        TA        TA            Mn            ALQ         348.0   

    bsmtfin_type_2  bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf heating  \
28             Unf           0.0        957.0          957.0    GasA   
53             Unf           0.0        370.0          546.0    GasA   
65             Unf           0.0        914.0         1120.0    GasA   
79             Unf           0.0        936.0          936.0    GasA   
101            Unf           0.0        180.0          528.0    GasA   

    heating_qc central_air electrical  1st_flr_sf  2nd_flr_sf  \
28          Fa           N      SBrkr        1034         957   
53          Ex           Y      SBrkr         546         546   
65          Ex           Y      SBrkr        1120           0   
79          Ex           Y      SBrkr         936         665   
101         Ex           Y      SBrkr         605           0   

     low_qual_fin_sf  gr_liv_area  bsmt_full_bath  bsmt_half_bath  full_bath  \
28                 0         1991             0.0             0.0          2   
53                 0         1092             0.0             0.0          1   
65                 0         1120             0.0             0.0          1   
79                 0         1601             0.0             0.0          2   
101                0   

In [34]:
# Identifying row of NaN values
print(train[train['garage_area'].isna()]['garage_area'])
print(train[train['garage_cars'].isna()]['garage_cars'])

1712   NaN
Name: garage_area, dtype: float64
1712   NaN
Name: garage_cars, dtype: float64


In [35]:
# Setting NaN values to 0 
train.loc[train[train['garage_area'].isna()].index, 'garage_area'] = 0.0
train.loc[train[train['garage_cars'].isna()].index, 'garage_cars'] = 0.0

In [36]:
# Verifying that the NaN values have been changed
print(train.groupby('garage_area').count()['id'].sum())
print(train.groupby('garage_cars').count()['id'].sum())

2051
2051


##### Cleaning column garage_yr_blt null values

In [37]:
len(train[train['garage_yr_blt'].isna()].index)

114

- As there is a small number of rows that are null values as compared to the entire training dataset.

- It would also not be practical to impute the null values by mean year per neighbourhood due to a possible inaccuracy and the mean year might possibly be less than the year the house was built.

- garage_yr_blt will be dropped for this case.

In [38]:
train.drop(index=train[train['garage_yr_blt'].isna()].index, inplace=True)

- Checking the rows left of train dataset when 114 rows are dropped from garage_yr_blt null values 

In [39]:
train.shape

(1937, 81)

- Training dataset now has 1937 rows

##### Cleaning column lot_frontage null values

In [40]:
# Identifying number of rows for NaN values
train[train['lot_frontage'].isna()]['lot_frontage'].shape

(324,)

In [41]:
# Setting Null values to 0 
train.loc[train[train['lot_frontage'].isna()].index, 'lot_frontage'] = 0.0

In [42]:
# Verifying that rows of the null values have been cleaned
train[train['lot_frontage'].isna()]['lot_frontage'].shape

(0,)

In [43]:
# Verifying that the null values has been changed
train.groupby('lot_frontage').count()['id'].sum()

1937

##### Cleaning column mas_vnr_area null values

In [44]:
# Identifying total num of rows for mas_vnr_area initially
train['mas_vnr_area'].shape

(1937,)

In [45]:
# Identifying total num of rows for mas_vnr_area with values initially
train.groupby('mas_vnr_area').count()['id'].sum()

1916

In [46]:
# Identifying number of rows for NA values initially
train[train['mas_vnr_area'].isna()].shape

(21, 81)

In [47]:
# Identifying number of '0' values initially
train[train['mas_vnr_area'] == 0].shape

(1114, 81)

In [48]:
# # Setting NaN values to 0 
train.loc[train[train['mas_vnr_area'].isna()].index, 'mas_vnr_area'] = 0.0

In [49]:
# Identifying number of '0' values after conversion
train[train['mas_vnr_area'] == 0].shape

(1135, 81)

Total 21 NaN values set to 0 has been verified and converted into column mas_vnr_area

In [50]:
# Verifying that the NaN values has been changed
train.groupby('mas_vnr_area').count()['id'].sum()

1937

##### Cleaning column total_bsmt_sf null values

In [51]:
# Identifying row of NaN values
train[train['total_bsmt_sf'].isna()]['total_bsmt_sf']

1327   NaN
Name: total_bsmt_sf, dtype: float64

In [52]:
# Setting NaN values to 0 
train.loc[train[train['total_bsmt_sf'].isna()].index, 'total_bsmt_sf'] = 0.0

In [53]:
# Verifying that the NaN values have been changed
train.groupby('total_bsmt_sf').count()['id'].sum()

1937

### Handling Categorical Null Values

In [54]:
#Getting the total null values for the Categorical columns in train dataset
train_null_cat = pd.DataFrame(train.isnull().sum(), columns=['null_count'])
train_null_cat = train_null_cat[train_null_cat['null_count'] > 0].sort_values('null_count', ascending=False).sort_index()
train_null_cat

null_count
alley                 1813
bsmt_cond               46
bsmt_exposure           49
bsmt_qual               46
bsmtfin_type_1          46
bsmtfin_type_2          47
fence                 1558
fireplace_qu           895
mas_vnr_type            21
misc_feature          1875
pool_qc               1928

##### Cleaning column alley null values

In [55]:
# Checking non null values
train['alley'].value_counts()

Grvl    71
Pave    53
Name: alley, dtype: int64

In [56]:
# Checking total null values
train['alley'].isnull().sum()

1813

In [57]:
# Percentage of null values for alley in train dataset
(train['alley'].isnull().sum())/2051*100

88.39590443686008

- As there is a total of 88.3% null values in Alley which is a large amount of null data, this possibily not likely to affect the pricing of the house. Thus column Alley shall be dropped.

In [58]:
train.drop(columns = 'alley', inplace=True)

In [59]:
train.head()

id        pid  ms_subclass ms_zoning  lot_frontage  lot_area street  \
0  109  533352170           60        RL           0.0     13517   Pave   
1  544  531379050           60        RL          43.0     11492   Pave   
2  153  535304180           20        RL          68.0      7922   Pave   
3  318  916386060           60        RL          73.0      9802   Pave   
4  255  906425045           50        RL          82.0     14235   Pave   

  lot_shape land_contour utilities lot_config land_slope neighborhood  \
0       IR1          Lvl    AllPub    CulDSac        Gtl       Sawyer   
1       IR1          Lvl    AllPub    CulDSac        Gtl      SawyerW   
2       Reg          Lvl    AllPub     Inside        Gtl        NAmes   
3       Reg          Lvl    AllPub     Inside        Gtl       Timber   
4       IR1          Lvl    AllPub     Inside        Gtl      SawyerW   

  condition_1 condition_2 bldg_type house_style  overall_qual  overall_cond  \
0        RRAe        Norm      1Fam      2Story             6             8   
1        Norm        Norm      1Fam      2Story             7             5   
2        Norm        Norm      1Fam      1Story             5             7   
3        Norm        Norm      1Fam      2Story             5             5   
4        Norm        Norm      1Fam      1.5Fin             6             8   

   year_built  year_remod/add roof_style roof_matl exterior_1st exterior_2nd  \
0        1976            2005      Gable   CompShg      HdBoard      Plywood   
1        1996            1997      Gable   CompShg      VinylSd      VinylSd   
2        1953            2007      Gable   CompShg      VinylSd      VinylSd   
3        2006            2007      Gable   CompShg      VinylSd      VinylSd   
4        1900            1993      Gable   CompShg      Wd Sdng      Plywood   

  mas_vnr_type  mas_vnr_area exter_qual exter_cond foundation bsmt_qual  \
0      BrkFace         289.0         Gd         TA     CBlock        TA   
1      BrkFace         132.0         Gd         TA      PConc        Gd   
2         None           0.0         TA         Gd     CBlock        TA   
3         None           0.0         TA         TA      PConc        Gd   
4         None           0.0         TA         TA      PConc        Fa   

  bsmt_cond bsmt_exposure bsmtfin_type_1  bsmtfin_sf_1 bsmtfin_type_2  \
0        TA            No            GLQ         533.0            Unf   
1        TA            No            GLQ         637.0            Unf   
2        TA            No            GLQ         731.0            Unf   
3        TA            No            Unf           0.0            Unf   
4        Gd            No            Unf           0.0            Unf   

   bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf heating heating_qc central_air  \
0           0.0        192.0          725.0    GasA         Ex           Y   
1           0.0        276.0          913.0    GasA         Ex           Y   
2           0.0        326.0         1057.0    GasA         TA           Y   
3           0.0        384.0          384.0    GasA         Gd           Y   
4           0.0        676.0          676.0    GasA         TA           Y   

  electrical  1st_flr_sf  2nd_flr_sf  low_qual_fin_sf  gr_liv_area  \
0      SBrkr         725         754                0         1479   
1      SBrkr         913        1209                0         2122   
2      SBrkr        1057           0                0         1057   
3      SBrkr         744         700                0         1444   
4      SBrkr         831         614                0         1445   

   bsmt_full_bath  bsmt_half_bath  full_bath  half_bath  bedroom_abvgr  \
0             0.0             0.0          2          1              3   
1             1.0             0.0          2          1              4   
2             1.0             0.0          1          0              3   
3             0.0             0.0          2          1              3   
4       

##### Cleaning column bsmt_cond null values

In [60]:
# Checking total non null values initially
train['bsmt_cond'].value_counts().sum()

1891

In [61]:
# Checking breakdown of non null values initially
train['bsmt_cond'].value_counts()

TA    1749
Gd      84
Fa      53
Ex       3
Po       2
Name: bsmt_cond, dtype: int64

In [62]:
# Checking total null values
train['bsmt_cond'].isnull().sum()

46

In [63]:
train[train['bsmt_cond'].isnull()].head()

id        pid  ms_subclass ms_zoning  lot_frontage  lot_area street  \
12    807  906226060           70        RL         120.0     26400   Pave   
93    811  906226100           90        RL          64.0      7018   Pave   
146   888  908128060           85        RL          64.0      7301   Pave   
183  1554  910251050           20   A (agr)          80.0     14584   Pave   
240  2740  905451050           20        RL          80.0     12048   Pave   

    lot_shape land_contour utilities lot_config land_slope neighborhood  \
12        Reg          Bnk    AllPub        FR2        Gtl      SawyerW   
93        Reg          Bnk    AllPub     Inside        Gtl      SawyerW   
146       Reg          Lvl    AllPub     Corner        Gtl      Edwards   
183       Reg          Low    AllPub     Inside        Mod       IDOTRR   
240       Reg          Lvl    AllPub     Inside        Gtl      Edwards   

    condition_1 condition_2 bldg_type house_style  overall_qual  overall_cond  \
12        Feedr        Norm      1Fam      2Story             5             7   
93         Norm        Norm    Duplex      1Story             5             5   
146        Norm        Norm      1Fam      SFoyer             7             5   
183        Norm        Norm      1Fam      1Story             1             5   
240        Norm        Norm      1Fam      1Story             5             6   

     year_built  year_remod/add roof_style roof_matl exterior_1st  \
12         1880            2007      Gable   CompShg      HdBoard   
93         1979            1979      Gable   CompShg      HdBoard   
146        2003            2003      Gable   CompShg      HdBoard   
183        1952            1952      Gable   CompShg      AsbShng   
240        1952            2002      Gable   CompShg      Wd Sdng   

    exterior_2nd mas_vnr_type  mas_vnr_area exter_qual exter_cond foundation  \
12       HdBoard         None           0.0         Gd         TA      PConc   
93       HdBoard         None           0.0         TA         TA       Slab   
146      HdBoard      BrkFace         500.0         Gd         TA       Slab   
183      VinylSd         None           0.0         Fa         Po       Slab   
240      Wd Sdng      BrkFace         232.0         TA         TA       Slab   

    bsmt_qual bsmt_cond bsmt_exposure bsmtfin_type_1  bsmtfin_sf_1  \
12        NaN       NaN           NaN            NaN           0.0   
93        NaN       NaN           NaN            NaN           0.0   
146       NaN       NaN           NaN            NaN           0.0   
183       NaN       NaN           NaN            NaN           0.0   
240       NaN       NaN           NaN            NaN           0.0   

    bsmtfin_type_2  bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf heating  \
12             NaN           0.0          0.0            0.0    GasA   
93             NaN           0.0          0.0            0.0    GasA   
146            NaN           0.0          0.0            0.0    GasA   
183            NaN           0.0          0.0            0.0    Wall   
240            NaN           0.0          0.0            0.0    GasA   

    heating_qc central_air electrical  1st_flr_sf  2nd_flr_sf  \
12          Ex           Y      SBrkr        1288         728   
93          TA           Y      SBrkr        1535           0   
146         Ex           Y      SBrkr         495        1427   
183         Po           N      FuseA         733           0   
240         Gd           Y      SBrkr        1488           0   

     low_qual_fin_sf  gr_liv_area  bsmt_full_bath  bsmt_half_bath  full_bath  \
12                 0         2016             0.0             0.0          1   
93                 0         1535             0.0             0.0          2   
146                0         1922             0.0             0.0          3   
183                0          733             0.0             0.0          1   
240                0         1488             0.0          

In [64]:
#Setting null values to 'None' to be an included value for categorical train dataset
train.loc[train[train['bsmt_cond'].isnull()].index, 'bsmt_cond'] = 'None'

In [65]:
# Checking updated total null values
train['bsmt_cond'].isnull().sum()

0

In [66]:
# Checking the null values are in the dataset as value counts
train['bsmt_cond'].value_counts()

TA      1749
Gd        84
Fa        53
None      46
Ex         3
Po         2
Name: bsmt_cond, dtype: int64

In [67]:
# Checking updated total non null values
train['bsmt_cond'].value_counts().sum()

1937

##### Cleaning column bsmt_exposure null values

In [68]:
# Checking total non null values initially
train['bsmt_exposure'].value_counts().sum()

1888

In [69]:
# Checking breakdown of non null values initially
train['bsmt_exposure'].value_counts()

No    1259
Av     276
Gd     195
Mn     158
Name: bsmt_exposure, dtype: int64

In [70]:
# Checking total null values
train['bsmt_exposure'].isnull().sum()

49

In [71]:
#Setting null values to 'None' to be an included value for categorical train dataset
train.loc[train[train['bsmt_exposure'].isnull()].index, 'bsmt_exposure'] = 'None'

In [72]:
# Checking updated total null values
train['bsmt_exposure'].isnull().sum()

0

In [73]:
# Checking the null values are in the dataset as value counts
train['bsmt_exposure'].value_counts()

No      1259
Av       276
Gd       195
Mn       158
None      49
Name: bsmt_exposure, dtype: int64

In [74]:
# Checking updated total non null values
train['bsmt_exposure'].value_counts().sum()

1937

##### Cleaning column bsmt_qual null values

In [75]:
# Checking total non null values initially
train['bsmt_qual'].value_counts().sum()

1891

In [76]:
# Checking breakdown of non null values initially
train['bsmt_qual'].value_counts()

Gd    839
TA    817
Ex    184
Fa     50
Po      1
Name: bsmt_qual, dtype: int64

In [77]:
# Checking total null values
train['bsmt_qual'].isnull().sum()

46

In [78]:
#Setting null values to 'None' to be an included value for categorical train dataset
train.loc[train[train['bsmt_qual'].isnull()].index, 'bsmt_qual'] = 'None'

In [79]:
# Checking updated total null values
train['bsmt_qual'].isnull().sum()

0

In [80]:
# Checking the null values are in the dataset as value counts
train['bsmt_qual'].value_counts()

Gd      839
TA      817
Ex      184
Fa       50
None     46
Po        1
Name: bsmt_qual, dtype: int64

In [81]:
# Checking updated total non null values
train['bsmt_qual'].value_counts().sum()

1937

##### Cleaning column bsmtfin_type_1 null values

In [82]:
# Checking total non null values initially
train['bsmtfin_type_1'].value_counts().sum()

1891

In [83]:
# Checking breakdown of non null values initially
train['bsmtfin_type_1'].value_counts()

GLQ    599
Unf    544
ALQ    279
BLQ    194
Rec    177
LwQ     98
Name: bsmtfin_type_1, dtype: int64

In [84]:
# Checking total null values
train['bsmtfin_type_1'].isnull().sum()

46

In [85]:
#Setting null values to 'None' to be an included value for categorical train dataset
train.loc[train[train['bsmtfin_type_1'].isnull()].index, 'bsmtfin_type_1'] = 'None'

In [86]:
# Checking updated total null values
train['bsmtfin_type_1'].isnull().sum()

0

In [87]:
# Checking the null values are in the dataset as value counts
train['bsmtfin_type_1'].value_counts()

GLQ     599
Unf     544
ALQ     279
BLQ     194
Rec     177
LwQ      98
None     46
Name: bsmtfin_type_1, dtype: int64

In [88]:
# Checking updated total non null values
train['bsmtfin_type_1'].value_counts().sum()

1937

##### Cleaning column bsmtfin_type_2 null values

In [89]:
# Checking total non null values initially
train['bsmtfin_type_2'].value_counts().sum()

1890

In [90]:
# Checking breakdown of non null values initially
train['bsmtfin_type_2'].value_counts()

Unf    1648
Rec      79
LwQ      60
BLQ      47
ALQ      34
GLQ      22
Name: bsmtfin_type_2, dtype: int64

In [91]:
# Checking total null values
train['bsmtfin_type_2'].isnull().sum()

47

In [92]:
#Setting null values to 'None' to be an included value for categorical train dataset
train.loc[train[train['bsmtfin_type_2'].isnull()].index, 'bsmtfin_type_2'] = 'None'

In [93]:
# Checking updated total null values
train['bsmtfin_type_2'].isnull().sum()

0

In [94]:
# Checking the null values are in the dataset as value counts
train['bsmtfin_type_2'].value_counts()

Unf     1648
Rec       79
LwQ       60
None      47
BLQ       47
ALQ       34
GLQ       22
Name: bsmtfin_type_2, dtype: int64

In [95]:
# Checking updated total non null values
train['bsmtfin_type_2'].value_counts().sum()

1937

##### Cleaning column fence null values

In [96]:
# Checking total non null values initially
train['fence'].value_counts().sum()

379

In [97]:
# Checking breakdown of non null values initially
train['fence'].value_counts()

MnPrv    214
GdPrv     82
GdWo      73
MnWw      10
Name: fence, dtype: int64

In [98]:
# Checking total null values
train['fence'].isnull().sum()

1558

In [99]:
# Updating exterior column reference as alley column is dropped
ext_column = ['id', 'pid', 'street', 'exterior_1st', 'exterior_2nd', 'mas_vnr_type','mas_vnr_area', 'exter_qual',
           'exter_cond', 'foundation', 'neighborhood','condition_1', 'condition_2', 'paved_drive', 'pool_qc','fence']

- Cross referencing with fields related to exteriors

In [100]:
train[ext_column].head()

id        pid street exterior_1st exterior_2nd mas_vnr_type  mas_vnr_area  \
0  109  533352170   Pave      HdBoard      Plywood      BrkFace         289.0   
1  544  531379050   Pave      VinylSd      VinylSd      BrkFace         132.0   
2  153  535304180   Pave      VinylSd      VinylSd         None           0.0   
3  318  916386060   Pave      VinylSd      VinylSd         None           0.0   
4  255  906425045   Pave      Wd Sdng      Plywood         None           0.0   

  exter_qual exter_cond foundation neighborhood condition_1 condition_2  \
0         Gd         TA     CBlock       Sawyer        RRAe        Norm   
1         Gd         TA      PConc      SawyerW        Norm        Norm   
2         TA         Gd     CBlock        NAmes        Norm        Norm   
3         TA         TA      PConc       Timber        Norm        Norm   
4         TA         TA      PConc      SawyerW        Norm        Norm   

  paved_drive pool_qc fence  
0           Y     NaN   NaN  
1           Y     NaN   NaN  
2           Y     NaN   NaN  
3           Y     NaN   NaN  
4           N     NaN   NaN

- Although it seems that column Fence has many NaN values, there is a possibility that it might be a valuable predictor for pricing. 


- Thus, assigning the NaN values as a neutral category for further investigation

In [101]:
#Setting null values to 'None' to be an included value for categorical train dataset
train.loc[train[train['fence'].isnull()].index, 'fence'] = 'None'

In [102]:
# Checking updated total null values
train['fence'].isnull().sum()

0

In [103]:
# Checking the null values are in the dataset as value counts
train['fence'].value_counts()

None     1558
MnPrv     214
GdPrv      82
GdWo       73
MnWw       10
Name: fence, dtype: int64

In [104]:
# Checking updated total non null values
train['bsmtfin_type_2'].value_counts().sum()

1937

##### Cleaning column fireplace_qu null values

In [105]:
# Checking total non null values initially
train['fireplace_qu'].value_counts().sum()

1042

In [106]:
# Checking breakdown of non null values initially
train['fireplace_qu'].value_counts()

Gd    517
TA    405
Fa     59
Ex     31
Po     30
Name: fireplace_qu, dtype: int64

In [107]:
# Checking total null values
train['fireplace_qu'].isnull().sum()

895

- Checking fireplace_qu & fireplaces if they share the same rows

In [108]:
train[train['fireplace_qu'].isna()].shape

(895, 80)

In [109]:
train[train['fireplaces'] == 0].shape

(895, 80)

In [110]:
# Checking for relationship for fireplace_qu
train[int_column].head()

id        pid  bsmt_full_bath  bsmt_half_bath  full_bath  half_bath  \
0  109  533352170             0.0             0.0          2          1   
1  544  531379050             1.0             0.0          2          1   
2  153  535304180             1.0             0.0          1          0   
3  318  916386060             0.0             0.0          2          1   
4  255  906425045             0.0             0.0          2          0   

   bedroom_abvgr  kitchen_abvgr kitchen_qual  totrms_abvgrd bsmt_qual  \
0              3              1           Gd              6        TA   
1              4              1           Gd              8        Gd   
2              3              1           Gd              5        TA   
3              3              1           TA              7        Gd   
4              3              1           TA              6        Fa   

  bsmt_cond bsmt_exposure bsmtfin_type_1  bsmtfin_sf_1 bsmtfin_type_2  \
0        TA            No            GLQ         533.0            Unf   
1        TA            No            GLQ         637.0            Unf   
2        TA            No            GLQ         731.0            Unf   
3        TA            No            Unf           0.0            Unf   
4        Gd            No            Unf           0.0            Unf   

   bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf  fireplaces fireplace_qu  \
0           0.0        192.0          725.0           0          NaN   
1           0.0        276.0          913.0           1           TA   
2           0.0        326.0         1057.0           0          NaN   
3           0.0        384.0          384.0           0          NaN   
4           0.0        676.0          676.0           0          NaN   

  garage_type  garage_yr_blt garage_finish  garage_cars  garage_area  \
0      Attchd         1976.0           RFn          2.0        475.0   
1      Attchd         1997.0           RFn          2.0        559.0   
2      Detchd         1953.0           Unf          1.0        246.0   
3     BuiltIn         2007.0           Fin          2.0        400.0   
4      Detchd         1957.0           Unf          2.0        484.0   

  garage_qual garage_cond  
0          TA          TA  
1          TA          TA  
2          TA          TA  
3          TA          TA  
4          TA          TA

- It may seem there is a possible chance that fireplace_qu might be a valuable predictor for pricing since it also matches the same number of rows with fireplaces = 0.


- Thus, assigning the NaN values as a neutral category for further investigation

In [111]:
#Setting null values to 'None' to be an included value for categorical train dataset
train.loc[train[train['fireplace_qu'].isnull()].index, 'fireplace_qu'] = 'None'

In [112]:
# Checking updated total null values
train['fireplace_qu'].isnull().sum()

0

In [113]:
# Checking the null values are in the dataset as value counts
train['fireplace_qu'].value_counts()

None    895
Gd      517
TA      405
Fa       59
Ex       31
Po       30
Name: fireplace_qu, dtype: int64

In [114]:
# Checking updated total non null values
train['fireplace_qu'].value_counts().sum()

1937

##### Cleaning column mas_vnr_type null values

In [115]:
# Checking total non null values initially
train['mas_vnr_type'].value_counts().sum()

1916

In [116]:
# Checking breakdown of non null values initially
train['mas_vnr_type'].value_counts()

None       1116
BrkFace     620
Stone       168
BrkCmn       12
Name: mas_vnr_type, dtype: int64

In [117]:
# Checking total null values
train['mas_vnr_type'].isnull().sum()

21

- Checking mas_vnr_type & mas_vnr_area if they share the same rows

In [118]:
train[train['mas_vnr_type'].isna() & (train['mas_vnr_area'] == 0)].shape

(21, 80)

- There is also a likelyhood that mas_vnr_type might be a valuable predictor for pricing since it also it matches the same number of rows as mas_vnr_area when mas_vnr_area = 0.


- Thus, assigning the NaN values for mas_vnr_type as a neutral category

In [119]:
#Setting null values to 'None' to be an included value for categorical train dataset
train.loc[train[train['mas_vnr_type'].isnull()].index, 'mas_vnr_type'] = 'None'

In [120]:
# Checking updated total null values
train['mas_vnr_type'].isnull().sum()

0

In [121]:
# Checking the null values are in the dataset as value counts
train['mas_vnr_type'].value_counts()

None       1137
BrkFace     620
Stone       168
BrkCmn       12
Name: mas_vnr_type, dtype: int64

In [122]:
# Checking updated total non null values
train['mas_vnr_type'].value_counts().sum()

1937

##### Cleaning column misc_feature null values

In [123]:
# Checking total non null values initially
train['misc_feature'].value_counts().sum()

62

In [124]:
# Checking breakdown of non null values initially
train['misc_feature'].value_counts()

Shed    54
Gar2     4
Othr     2
Elev     1
TenC     1
Name: misc_feature, dtype: int64

In [125]:
# Checking total null values
train['misc_feature'].isnull().sum()

1875

In [126]:
# Checking relationship for misc_feature in related utility fields
train[util_column].head()

id        pid utilities heating heating_qc central_air electrical  \
0  109  533352170    AllPub    GasA         Ex           Y      SBrkr   
1  544  531379050    AllPub    GasA         Ex           Y      SBrkr   
2  153  535304180    AllPub    GasA         TA           Y      SBrkr   
3  318  916386060    AllPub    GasA         Gd           Y      SBrkr   
4  255  906425045    AllPub    GasA         TA           Y      SBrkr   

  misc_feature  misc_val  
0          NaN         0  
1          NaN         0  
2          NaN         0  
3          NaN         0  
4          NaN         0

- A possibility for misc_feature to be a valuable predictor for pricing, which will be used for further investigation.


- Thus, assigning the NaN values for misc_feature as a neutral category

In [127]:
#Setting null values to 'None' to be an included value for categorical train dataset
train.loc[train[train['misc_feature'].isnull()].index, 'misc_feature'] = 'None'

In [128]:
# Checking updated total null values
train['misc_feature'].isnull().sum()

0

In [129]:
# Checking the null values are in the dataset as value counts
train['misc_feature'].value_counts()

None    1875
Shed      54
Gar2       4
Othr       2
Elev       1
TenC       1
Name: misc_feature, dtype: int64

In [130]:
# Checking updated total non null values
train['misc_feature'].value_counts().sum()

1937

##### Cleaning column pool_qc null values

In [131]:
# Checking total non null values initially
train['pool_qc'].value_counts().sum()

9

In [132]:
# Checking breakdown of non null values initially
train['pool_qc'].value_counts()

Gd    4
Fa    2
TA    2
Ex    1
Name: pool_qc, dtype: int64

In [133]:
# Checking total null values
train['pool_qc'].isnull().sum()

1928

In [134]:
# Checking relationship for pool_qc in related utility fields
train[qc_column].head()

id        pid  overall_qual  overall_cond exter_qual exter_cond pool_qc  \
0  109  533352170             6             8         Gd         TA     NaN   
1  544  531379050             7             5         Gd         TA     NaN   
2  153  535304180             5             7         TA         Gd     NaN   
3  318  916386060             5             5         TA         TA     NaN   
4  255  906425045             6             8         TA         TA     NaN   

  kitchen_qual bsmt_qual bsmt_cond fireplace_qu garage_qual garage_cond  \
0           Gd        TA        TA         None          TA          TA   
1           Gd        Gd        TA           TA          TA          TA   
2           Gd        TA        TA         None          TA          TA   
3           TA        Gd        TA         None          TA          TA   
4           TA        Fa        Gd         None          TA          TA   

  heating_qc  
0         Ex  
1         Ex  
2         TA  
3         Gd  
4         TA

- Another likelyhood for qc_column to be a valuable predictor for pricing, which will be used for further investigation.


- Thus, assigning the NaN values for qc_column as a neutral category

In [135]:
#Setting null values to 'None' to be an included value for categorical train dataset
train.loc[train[train['pool_qc'].isnull()].index, 'pool_qc'] = 'None'

In [136]:
# Checking updated total null values
train['pool_qc'].isnull().sum()

0

In [137]:
# Checking the null values are in the dataset as value counts
train['pool_qc'].value_counts()

None    1928
Gd         4
Fa         2
TA         2
Ex         1
Name: pool_qc, dtype: int64

In [138]:
# Checking updated total non null values
train['pool_qc'].value_counts().sum()

1937

#### Getting null values and checking if any null values are left within all columns for training dataset

In [139]:
# Getting the total null values for the all columns in train dataset
train_null = pd.DataFrame(train.isnull().sum(), columns = ['null_count']).sort_values(by = 'null_count',ascending = False)
train_null = train_null[train_null['null_count'] > 0].sort_values('null_count', ascending=False).sort_index()

# Checking there are no null values
train_null

Empty DataFrame
Columns: [null_count]
Index: []

- Checking if each training columns values matching the total number of rows

In [140]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1937 entries, 0 to 2050
Data columns (total 80 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               1937 non-null   int64  
 1   pid              1937 non-null   int64  
 2   ms_subclass      1937 non-null   int64  
 3   ms_zoning        1937 non-null   object 
 4   lot_frontage     1937 non-null   float64
 5   lot_area         1937 non-null   int64  
 6   street           1937 non-null   object 
 7   lot_shape        1937 non-null   object 
 8   land_contour     1937 non-null   object 
 9   utilities        1937 non-null   object 
 10  lot_config       1937 non-null   object 
 11  land_slope       1937 non-null   object 
 12  neighborhood     1937 non-null   object 
 13  condition_1      1937 non-null   object 
 14  condition_2      1937 non-null   object 
 15  bldg_type        1937 non-null   object 
 16  house_style      1937 non-null   object 
 17  overall_qual  

- It appears that ms_zoning has some seperated spacing in its categorical naming from the data dictionary

In [141]:
train['ms_zoning'].value_counts()

RL         1531
RM          280
FV          101
RH           11
C (all)      11
A (agr)       2
I (all)       1
Name: ms_zoning, dtype: int64

- Replacing categorical naming of ms_zoning with seperated spacing to a single character for easy referencing

In [142]:
train['ms_zoning'].replace({'C (all)': 'C', 'A (agr)': 'A', 'I (all)': 'I'}, inplace=True)

In [143]:
train['ms_zoning'].value_counts()

RL    1531
RM     280
FV     101
C       11
RH      11
A        2
I        1
Name: ms_zoning, dtype: int64

#### Writing cleaned training dataset into a csv file to datasets folder

In [144]:
train.to_csv('../datasets/train_clean.csv', index=False)